ext/memcached_rb/memcached_rb.c

```diff
@@ -1,460 +1,518 @@
 
-  return (rc == MEMCACHED_SUCCESS);
+	return (rc == MEMCACHED_SUCCESS) ? Qtrue : Qfalse;
 }
 
 static VALUE
-rb_connection_set(VALUE self, VALUE key, VALUE value, VALUE ttl, VALUE flags)
+rb_connection_set(VALUE self, VALUE rb_key, VALUE rb_value, VALUE rb_ttl, VALUE rb_flags)
 {
-  memcached_ctx *ctx = get_ctx(self);
-
-  char *mkey = StringValuePtr(key);
-  char *mvalue = StringValuePtr(value);
-
-  memcached_return_t rc = memcached_set(ctx->memc, mkey, strlen(mkey), mvalue, strlen(mvalue), NUM2INT(ttl), NUM2INT(flags));
-  Wrap_return_t(rc);
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
+	Check_Type(rb_value, T_STRING);
+	Check_Type(rb_ttl, T_FIXNUM);
+	Check_Type(rb_flags, T_FIXNUM);
+
+	rc = memcached_set(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		RSTRING_PTR(rb_value), RSTRING_LEN(rb_value),
+		FIX2INT(rb_ttl), FIX2INT(rb_flags)
+	);
+			
+	rb_memcached_return(rc);
 }
 
 static VALUE
-rb_connection_get(VALUE self, VALUE key)
+rb_connection_get(VALUE self, VALUE rb_key)
 {
-  Check_Type(key, T_STRING);
+	VALUE rb_val;
+
+	memcached_st *mc;
+	memcached_return_t rc;
+	size_t ret_len;
+	uint32_t ret_flags;
+	char *ret;
 
-  memcached_ctx *ctx = get_ctx(self);
+	UnwrapMemcached(self, mc);
+	Check_Type(rb_key, T_STRING);
 
-  char *mkey = StringValuePtr(key);
+	ret = memcached_get(mc,
+		RSTRING_PTR(rb_key), RSTRING_LEN(rb_key),
+		&ret_len, &ret_flags, &rc);
 
-  size_t return_value_length;
-  memcached_return_t rc;
-  uint32_t flags; // TODO return flags so client can know if it should deserialize or decompress
-  char *response  = memcached_get(ctx->memc, mkey, strlen(mkey), &return_value_length, &flags, &rc);
+	if (ret == NULL) {
+		rb_memcached_error_check(rc);
+		return Qnil;
+	}
 
-  handle_memcached_return(rc);
-  if (rc == MEMCACHED_SUCCESS)
-    return rb_str_new2(response);
-  else
-    return Qnil;
+	rb_val = rb_str_new(ret, ret_len);
+	free(ret);
+
+	/* TODO: return the flags too, maybe */
+	return rb_val;
+}
+
+memcached_return_t
+rb_connection__mget_callback(const memcached_st *mc, memcached_result_st *result, void *context)
+{
+	VALUE rb_result = (VALUE)context;
+	VALUE rb_key = rb_str_new(
+			memcached_result_key_value(result),
+			memcached_result_key_length(result));
+	VALUE rb_value = rb_str_new(
+			memcached_result_value(result),
+			memcached_result_length(result));
+
+	rb_hash_aset(rb_result, rb_key, rb_value);
+	return MEMCACHED_SUCCESS;
 }
 
 static VALUE
 rb_connection_get_multi(VALUE self, VALUE rb_keys)
 {
-  memcached_ctx *ctx = get_ctx(self);
-  VALUE rb_values = rb_hash_new();
-  int i;
-  long keys_len = RARRAY_LEN(rb_keys);
-  const char * keys[keys_len];
-  size_t key_length[keys_len];
+	memcached_execute_fn callbacks[1] = { &rb_connection__mget_callback };
+	memcached_st *mc;
+	memcached_return_t rc;
+
+	long i, keys_len;
+	const char **c_keys;
+	size_t *c_lengths;
+
+	VALUE rb_result;
 
-  VALUE * arr = RARRAY_PTR(rb_keys);
```

Comment:
`RARRAY_PTR` is an evil API. It will not work on Rubinius. Let's use `rb_ary_entry` instead, even though it can be a tad slower.